### PR TITLE
Handle failure during thread creation

### PR DESCRIPTION
**Pull request by antonpirker**
In Python 3.12 when you try to start a thread during shutdown a `RunTimeErrror` is raised. Handle this case with grace

### CODE REVIEW

File: sentry_sdk/profiler.py [hunk 1]
```diff
@@ -918,7 +918,13 @@ def ensure_running(self):
...
             # have exited
             self.thread = threading.Thread(name=self.name, target=self.run, daemon=True)
-            self.thread.start()
+            try:
+                self.thread.start()
+            except RuntimeError:
+                # Unfortunately at this point the interpreter is in a state that no
+                # longer allows us to spawn a thread and we have to bail.
+                self.running = False
+                return
```
Comment (szokeasaurusrex) [inline sentry_sdk/profiler.py:927]: I find it slightly confusing here that the method is called `ensure_running`, but it is possible for the method to terminate without the thread running. We should have some way to indicate failure, perhaps by raising a custom exception when `ensure_running` cannot start the scheduler or by returning a value to indicate success/failure. We might instead or additionally consider renaming the function to indicate that we may fail to ensure that the scheduler is running and/or add a documentation comment.

If we make this change to indicate when the `ensure_running` fails, we should also make sure to handle failure appropriately in the calling code.

File: sentry_sdk/sessions.py [hunk 2]
```diff
@@ -120,9 +120,17 @@ def _thread():
...
             thread = Thread(target=_thread)
             thread.daemon = True
-            thread.start()
+            try:
+                thread.start()
+            except RuntimeError:
+                # Unfortunately at this point the interpreter is in a state that no
+                # longer allows us to spawn a thread and we have to bail.
+                self._running = False
+                return None
```
Comment (szokeasaurusrex) [inline sentry_sdk/sessions.py:129]: See my comment on `profiler.py`

File: sentry_sdk/monitor.py [hunk 3]
```diff
@@ -53,7 +53,14 @@ def _thread():
...
             thread = Thread(name=self.name, target=_thread)
             thread.daemon = True
-            thread.start()
+            try:
+                thread.start()
+            except RuntimeError:
+                # Unfortunately at this point the interpreter is in a state that no
+                # longer allows us to spawn a thread and we have to bail.
+                self._running = False
+                return None
```
Comment (szokeasaurusrex) [inline sentry_sdk/monitor.py:62]: See my comment on `profiler.py`

**Review by szokeasaurusrex — COMMENTED** [review]
Please see my comment on `profiler.py`; it applies to all the files.

I also noticed an `ensure_running` function in the `GeventScheduler` of `profiler.py` that was unchanged in this PR, even though we are spawning a thread there. Does this method also need to be updated, or is a `RuntimeError` impossible here?

If possible, we might also add some tests to ensure this behavior works correctly

**Review by sl0thentr0py — APPROVED** [review]

File: sentry_sdk/sessions.py (hunk 2 above)
Comment (sentrivana) [reply sentry_sdk/sessions.py:136]: Followed up there

File: sentry_sdk/monitor.py (hunk 3 above)
Comment (sentrivana) [reply sentry_sdk/monitor.py:69]: Followed up there

**Review by szokeasaurusrex — APPROVED** [review]
I approve

**Pull request merged**